integration-templates/sharepoint-online/syncs/file-sync.ts

```diff
@@ -1,121 +1,108 @@
-
-        const documentLibraries = lists.filter((list: any) => list.list.template === 'documentLibrary');
-        siteIdToLists[siteId] = documentLibraries.map((l: any) => l.id);
+    for (const site of sitesToSync) {
+        const siteId = site.id;
+        const config = {
+            endpoint: `v1.0/sites/${siteId}/lists`,
+            paginate: {
+                type: 'link',
+                limit_name_in_request: '$top',
+                response_path: 'value',
+                link_path_in_response_body: '@odata.nextLink',
+                limit: 100
+            },
+            retries: 10
+        };
+        // Paginate through lists and filter documentlibraries
+        for await (const lists of nango.paginate(config)) {
+            const documentLibraries = lists.filter((list: any) => list.list.template === 'documentLibrary');
+            siteIdToLists[siteId] = documentLibraries.map((l: any) => l.id);
+        }
     }
 
     return siteIdToLists;
 }
 
-function getDeltaToken(metadata: SharePointMetadata, siteId: string, listId: string): string {
-    return metadata.syncedSites?.find((site) => site.id === siteId && site.listId === listId)?.deltaToken || '';
-}
-
-async function processListItems(nango: NangoSync, params: PaginationParams, siteId: string, listId: string): Promise<string> {
-    let newDeltaToken = '';
+/**
+ * Processes list items for synchronization from a SharePoint list.
+ *
+ * @param nango An instance of NangoSync for handling synchronization tasks.
+ * @param siteId The ID of the SharePoint site containing the list.
+ * @param listId The ID of the SharePoint list containing items to sync.
+ * @returns Promise<void>
+ */
+async function processListItems(nango: NangoSync, siteId: string, listId: string): Promise<void> {
+    const config = {
+        endpoint: `/v1.0/sites/${siteId}/lists/${listId}/items/delta`,
+        paginate: {
+            type: 'link',
+            limit_name_in_request: '$top',
+            response_path: 'value',
+            link_path_in_response_body: '@odata.nextLink',
+            limit: 100
+        },
+        // Include '$filter' parameter with the lastSyncDate if available
+        ...(nango.lastSyncDate ? { params: { $filter: `lastModifiedDateTime ge ${nango.lastSyncDate.toISOString()}` } } : {}),
+        retries: 10
+    };
 
-    await collectPaginatedData<FileMetadata>(nango, params, async (page: PaginationResponse<FileMetadata>) => {
-        for (const item of page.value) {
+    // Paginate through list items and sync each file metadata
+    for await (const listItems of nango.paginate(config)) {
+        for (const item of listItems) {
             const file = await fetchDriveItemDetails(nango, siteId, listId, item.id);
-            const metadata = fileMetadata(file);
-            await nango.batchSave<FileMetadata>([metadata], 'FileMetadata');
+            await nango.batchSave<FileMetadata>([file], 'FileMetadata');
```

Comment:
This implementation will introduce parallel requests, which can potentially defeat the request retry policy. Additionally, it may cause problems due to Sharepoint’s API throttling. Please let me know if we should proceed with this approach?

Reply:
good point about concurrency. we can still fetch the details synchronously and make a single call to batchSave instead of one per item. What do you think?
```
for await (const listItems of nango.paginate(config)) {
        const allMetadata = listItems.map((item) => await fetchDriveItemDetails(nango, siteId, listId, item.id));
        await nango.batchSave<FileMetadata>([allMetadata], 'FileMetadata');
}
```

Reply:
That is a better alternative. However, my initial focus was also on optimizing memory efficiency. Wouldn't this alternative approach potentially impact memory efficiency, especially when we hold data for too long, such as during retries while fetching drive item? What do you think?

Reply:
is memory a problem? also how big are pages? If they are a few hundred items I think it is fine to hold them in memory before to batchSave them. The overhead to batchSave (which is a round-trip to persist over the network) would make saving much slower.

Reply:
Yes, in my opinion, we need to consider memory efficiency when writing syncs. We should batch save as soon as possible. However, in this case, we can hold in memory since we're dealing with a slimmer file object and not saving the actual file itself. The response is paginated, so we will always have fewer records to batch save. I've also made a small tweak to the code: we cannot use `await` directly inside a synchronous callback function like `map`. Please take a look at the [changes](https://github.com/NangoHQ/nango/pull/2495/commits/3a0c1ce17b46a86e78475e0c3b2a0279c16e73f8) and provide me with your feedback. Thank you!